### DISCUSSION TITLE
Python debugging

### DISCUSSION THREAD

**Discussion by antocuni** [pyscript/pyscript#1597] [Proposals]
In this discussion I'd like to share some of my thoughts, ideas and future plans about debugging in PyScript.

Here, I am specifically talking about "debugging python apps written in pyscript". I am *not* talking about:
  - debugging pyscript itself
  - debugging other languages supported by pyscript

---
### Desirable features

The amount of features that you can have is endless. I think that for the very first iteration, we want at least:
1. the ability of setting a breakpoint
2. the ability of running code step-by-step, line-by-line, etc.
3. the ability of inspecting local and global variables
4. the ability of inspecting what happened in case of exception

---
### How debugging works in CPython

This is a brief explanation of how debugging work in CPython. The following info are true for all OS, including linux, win, mac. Nothing is specific to pyodide, wasm or the browser here.

The core of the functionality is offered by `sys.settrace()`.
By calling `sys.settrace()`, you install a "trace function" which is called whenever certain events occur, e.g. before executing any line of code, before calling any function, before returning, etc. This function can be implemented in Python or C.

Debuggers work by installing a trace function which keep tracks of the lines being executed. Normally, the trace function is called very frequently and returns immediately, but when it sees the `line` event on a line on which we set a breakpoint, the trace function simply doesn't return immediately. It **waits** until the user do something, e.g. pressing a button, and return only when it wants the execution to continue. This is how we implement the **pausing** of a debugged program.

The "wait" part is implemented using a **blocking syscall**, so that the OS suspends the execution of the program until something happens. The blocking syscall can be e.g. "read a line from the keyboard", "read a packet from the network" or "wait until I get this lock" (and presumably, the lock is released by another thread in which you have a GUI which responds to events).

Here is a super-simplified version of a fake debugger which always stops at line 7:
https://gist.github.com/antocuni/4920643fade57a03456f25c9a3450dde

<details>
  <summary>fakepdb.py</summary>

```python
import sys

def main():
    myvar = 'hello'
    print('this is line 5')
    print('this is line 6')
    print('this is line 7')
    print('this is line 8')
    print('this is line 9')

def fakepdb(frame, event, arg):
    if event == 'line':
        print('      about to execute line', frame.f_lineno)
        if frame.f_lineno == 7:
            print('welcome to fakepdb: type "c" to continue, or the name of '
                  'a variable to print it')
            while True:
                cmd = input('(fakepdb) ')
                if cmd == 'c':
                    break
                else:
                    v = frame.f_locals.get(cmd, f'variable {cmd} not found')
                    print(v)
    return fakepdb # continue tracing

if __name__ == '__main__':
    sys.settrace(fakepdb)
    main()
```

</details>

```
$ python fakepdb.py
      about to execute line 4
      about to execute line 5
this is line 5
      about to execute line 6
this is line 6
      about to execute line 7
welcome to fakepdb: type "c" to continue, or the name of a variable to print it
(fakepdb) myvar
hello
(fakepdb) xxx
variable xxx not found
(fakepdb) c
this is line 7
      about to execute line 8
this is line 8
      about to execute line 9
this is line 9
```

---
### Debugging in the browser

The key part is that python debuggers stop the execution of the program by calling (directly or indirectly) a **blocking syscall**.
In the fakepdb example above, we call `input()` which calls something something which eventually calls `read(stdin)`, which is the actual blocking syscall.

In the browser, we cannot block in the main thread.
As far as I know, python debugging is **impossible when pyodide runs in the main thread**. I cannot even imagine any good* workaround for this.

<details>
  <summary>Ugly workaround</summary>

I suppose that in theory if we asyncify the whole cpython it could work, but I don't think it's a good idea for many many reasons -- let's not waste time on this here.
  
</details>

However, the good news is that we can "easily" implement python debugging when pyodide runs in a worker: thanks to the new `xworker.sync` functionality which is coming in pyscript-next, we can easily have blocking calls that we can call from our custom tracing function to stop the execution.

---
### Pyscript debugging, step 0

Once we have:
1) pyscript-next
2) a `<py-terminal>` based on `xterm.js`
3) the `input()` function implemented using `xworker.sync` and xterm.js

I **think** that `pdb` will work out of the box.
This is surely not the "easy debugger" that we want to give to our users, but it is surely a first good step to check that all the machinery works as intended, and it will already be an incredible step forward compared to the current situation.

### PyScript debugging, step 1

Once we have step 0 working, we can start thinking of a better solution. I **think** it should be relatively easy to implement a basic GUI debugger. The core of the idea is:
1) in the worker, we will install a custom trace function similar to what I have shown above (possibly based on `bdb` from the stdlib)
2) in the main thread, we have a GUI which shows the source code, line numbers, buttons to resume the execution, etc.
3) the worker and the main thread communicate using the standard postmessage dance

### PyScript debugging, step 2

The "step 1" debugger described above shows the GUI in the DOM. In theory, it should be possible to setup things in such a way that instead of sending messages to the main thread, the worker sends messages to "something else": this would allow external programs to debug our pyscript app: for example, I can imagine something like this to debug an app from pyscript.com.

I have no idea how this interacts with security restriction of the browser, CORS, COEP, etc.

### PyScript debugging, step 2.5

There is a thing called [Debug Adapter Protocol (DAP)](https://microsoft.github.io/debug-adapter-protocol/). TL;DR: it's a protocol to allow IDEs to debug programs in a language-neutral way.
I believe it is spoke by vscode.

In theory, we could try to use DAP for step 2. I think there are already python debugging servers which speaks DAP, but I don't know how hard it is to run them in the browser. It is surely something that we need to investigate.

For all I know, there *might* be some DAP client implemented in JS that we could try to reuse/integrate, but again I didn't investigate.

---
### Bonus section: why we can't use the Chrome Dev Tools?

Because the dev tools operate at a fundamentally different level.
Pyodide is "CPython compiled to WASM", i.e. an interpreter.
From the dev tools point of view, you can put breakpoints on C-level functions such as `PyEval_EvalFrame` etc., but it is not what we want: *all* user-written python code is executed inside `PyEval_EvalFrame`, so it's not very useful.

It is basically equivalent to use `gdb` on linux: it is useful if you want to debug the C code inside CPython, not to debug your python program.

---
### What about MicroPython and other languages?

TL;DR: I have no idea how debugging works on those interpreters :)
